File: js/UI/DialogueSystem.js
// Dialogue System
class DialogueSystem {
    constructor(gameManager) {
        this.gameManager = gameManager;
        this.dialogueContainer = document.getElementById('dialogue-container');
        this.dialogueText = document.getElementById('dialogue-text');
        this.dialogueChoices = document.getElementById('dialogue-choices');
        this.speakerName = document.getElementById('speaker-name');
        this.speakerPortrait = document.querySelector('.speaker-portrait');
        
        this.onChoiceSelected = null;
        this.typewriterInterval = null;
        this.fullDialogueText = '';
        this.useTypewriterEffect = true;
        this.typewriterSpeed = 50; // ms per character
    }
    
    /**
     * Show dialogue with choices
     * @param {string} text - The dialogue text
     * @param {Array} choices - Array of choice texts
     * @param {Function} onChoiceSelected - Callback when choice is selected
     */
    showDialogue(text, choices, onChoiceSelected) {
        this.fullDialogueText = text;
        this.onChoiceSelected = onChoiceSelected;
        
        // Show the dialogue container
        if (this.dialogueContainer) {
            this.dialogueContainer.classList.remove('hidden');
        }
        
        // Hide choices initially while text is typing
        if (this.dialogueChoices) {
            clearChildren(this.dialogueChoices);
        }
        
        // Set dialogue text
        if (this.dialogueText) {
            if (this.useTypewriterEffect) {
                this.startTypewriterEffect(text, choices);
            } else {
                this.dialogueText.textContent = text;
                this.createChoiceButtons(choices);
            }
        }
    }
    
    /**
     * Show dialogue with a speaker
     * @param {string} text - The dialogue text
     * @param {Array} choices - Array of choice texts
     * @param {Function} onChoiceSelected - Callback when choice is selected
     * @param {string} speakerName - Name of the speaker
     * @param {string} portraitUrl - URL to speaker portrait
     */
    showDialogueWithSpeaker(text, choices, onChoiceSelected, speakerName, portraitUrl) {
        // Set speaker name
        if (this.speakerName) {
            this.speakerName.textContent = speakerName;
        }
        
        // Set portrait if available
        if (this.speakerPortrait) {
            if (portraitUrl) {
                this.speakerPortrait.style.backgroundImage = `url(${portraitUrl})`;
                this.speakerPortrait.style.display = 'block';
            } else {
                this.speakerPortrait.style.display = 'none';
            }
        }
        
        // Show regular dialogue
        this.showDialogue(text, choices, onChoiceSelected);
    }
    
    /**
     * Start typewriter effect for gradual text reveal
     * @param {string} text - The text to type
     * @param {Array} choices - The choices to show after text is typed
     */
    startTypewriterEffect(text, choices) {
        // Clear any existing typewriter interval
        if (this.typewriterInterval) {
            clearInterval(this.typewriterInterval);
        }
        
        let index = 0;
        this.dialogueText.textContent = '';
        
        // Start typing
        this.typewriterInterval = setInterval(() => {
            if (index < text.length) {
                this.dialogueText.textContent += text.charAt(index);
                index++;
            } else {
                // Text is complete, show choices
                clearInterval(this.typewriterInterval);
                this.typewriterInterval = null;
                this.createChoiceButtons(choices);
            }
        }, this.typewriterSpeed);
        
        // Add click listener to skip typing
        this.dialogueContainer.onclick = (e) => {
            // Only if we clicked the dialogue container itself, not a choice
            if (e.target === this.dialogueContainer || e.target === this.dialogueText) {
                if (this.typewriterInterval) {
                    // Complete the text immediately
                    clearInterval(this.typewriterInterval);
                    this.typewriterInterval = null;
                    this.dialogueText.textContent = text;
                    this.createChoiceButtons(choices);
                }
            }
        };
    }
    
    /**
     * Create choice buttons
     * @param {Array} choices - Array of choice texts
     */
    createChoiceButtons(choices) {
        if (!this.dialogueChoices) return;
        
        // Clear existing choices
        clearChildren(this.dialogueChoices);
        
        // Create a button for each choice
        choices.forEach((choice, index) => {
            const button = createElement('div', {
                className: 'dialogue-choice',
                textContent: choice,
                onClick: () => this.onChoiceButtonClicked(index)
            });
            
            this.dialogueChoices.appendChild(button);
        });
    }
    
    /**
     * Handle choice button clicked
     * @param {number} choiceIndex - The index of the choice
     */
    onChoiceButtonClicked(choiceIndex) {
        if (this.onChoiceSelected) {
            this.onChoiceSelected(choiceIndex);
        }
    }
    
    /**
     * Hide the dialogue
     */
    hideDialogue() {
        if (this.dialogueContainer) {
            this.dialogueContainer.classList.add('hidden');
        }
        
        // Clear any typewriter effect
        if (this.typewriterInterval) {
            clearInterval(this.typewriterInterval);
            this.typewriterInterval = null;
        }
    }
    
    /**
     * Complete the text typing immediately
     */
    completeTyping() {
        if (this.typewriterInterval) {
            clearInterval(this.typewriterInterval);
            this.typewriterInterval = null;
            
            if (this.dialogueText) {
                this.dialogueText.textContent = this.fullDialogueText;
            }
        }
    }
}

File: Assets/Scripts/UI/DialogueSystem.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;
using TMPro;

/// <summary>
/// Manages dialogue display and choice selection
/// </summary>
public class DialogueSystem : MonoBehaviour
{
    [SerializeField] private TMP_Text _dialogueText;
    [SerializeField] private Transform _choicesContainer;
    [SerializeField] private GameObject _choiceButtonPrefab;
    [SerializeField] private Image _speakerPortrait;
    [SerializeField] private TMP_Text _speakerNameText;
    [SerializeField] private Image _dialogueBackgroundPanel;
    [SerializeField] private float _typewriterSpeed = 0.05f;
    [SerializeField] private bool _useTypewriterEffect = true;
    
    private System.Action<int> _onChoiceSelected;
    private Coroutine _typewriterCoroutine;
    private string _fullDialogueText;
    
    private void Awake()
    {
        // Make sure the dialogue system is initially hidden
        gameObject.SetActive(false);
    }
    
    /// <summary>
    /// Display dialogue with choices
    /// </summary>
    public void ShowDialogue(string text, List<string> choices, System.Action<int> onChoiceSelected)
    {
        _fullDialogueText = text;
        _onChoiceSelected = onChoiceSelected;
        
        // Show the dialogue panel
        gameObject.SetActive(true);
        
        // Hide choices initially while text is typing
        if (_choicesContainer != null)
            _choicesContainer.gameObject.SetActive(false);
        
        // Set dialogue text
        if (_dialogueText != null)
        {
            if (_useTypewriterEffect)
            {
                if (_typewriterCoroutine != null)
                    StopCoroutine(_typewriterCoroutine);
                    
                _dialogueText.text = "";
                _typewriterCoroutine = StartCoroutine(TypewriterEffect(_fullDialogueText, choices));
            }
            else
            {
                _dialogueText.text = text;
                CreateChoiceButtons(choices);
            }
        }
    }
    
    /// <summary>
    /// Display dialogue with a speaker portrait
    /// </summary>
    public void ShowDialogueWithSpeaker(string text, List<string> choices, System.Action<int> onChoiceSelected, 
                                        string speakerName, Sprite portrait)
    {
        // Setup speaker elements
        if (_speakerNameText != null)
        {
            _speakerNameText.text = speakerName;
            _speakerNameText.gameObject.SetActive(true);
        }
        
        if (_speakerPortrait != null)
        {
            if (portrait != null)
            {
                _speakerPortrait.sprite = portrait;
                _speakerPortrait.gameObject.SetActive(true);
            }
            else
            {
                _speakerPortrait.gameObject.SetActive(false);
            }
        }
        
        // Show regular dialogue
        ShowDialogue(text, choices, onChoiceSelected);
    }
    
    /// <summary>
    /// Create buttons for each dialogue choice
    /// </summary>
    private void CreateChoiceButtons(List<string> choices)
    {
        if (_choicesContainer == null || _choiceButtonPrefab == null)
            return;
            
        // Clear existing choices
        foreach (Transform child in _choicesContainer)
        {
            Destroy(child.gameObject);
        }
        
        // Show choices container
        _choicesContainer.gameObject.SetActive(true);
        
        // Create a button for each choice
        for (int i = 0; i < choices.Count; i++)
        {
            GameObject choiceObj = Instantiate(_choiceButtonPrefab, _choicesContainer);
            
            // Set choice text
            TMP_Text choiceText = choiceObj.GetComponentInChildren<TMP_Text>();
            if (choiceText != null)
            {
                choiceText.text = choices[i];
            }
            
            // Add click event
            Button choiceButton = choiceObj.GetComponent<Button>();
            if (choiceButton != null)
            {
                int choiceIndex = i;
                choiceButton.onClick.AddListener(() => {
                    OnChoiceButtonClicked(choiceIndex);
                });
            }
        }
    }
    
    /// <summary>
    /// Handle choice button click
    /// </summary>
    private void OnChoiceButtonClicked(int choiceIndex)
    {
        if (_onChoiceSelected != null)
        {
            _onChoiceSelected.Invoke(choiceIndex);
        }
    }
    
    /// <summary>
    /// Typewriter effect for gradual text reveal
    /// </summary>
    private IEnumerator TypewriterEffect(string textToType, List<string> choices)
    {
        _dialogueText.text = "";
        
        foreach (char c in textToType)
        {
            _dialogueText.text += c;
            
            // Skip to end if player taps dialogue box
            if (Input.GetMouseButtonDown(0))
            {
                _dialogueText.text = textToType;
                break;
            }
            
            yield return new WaitForSeconds(_typewriterSpeed);
        }
        
        // Show choices after text is fully displayed
        CreateChoiceButtons(choices);
        
        _typewriterCoroutine = null;
    }
    
    /// <summary>
    /// Hide the dialogue system
    /// </summary>
    public void HideDialogue()
    {
        gameObject.SetActive(false);
        
        if (_typewriterCoroutine != null)
        {
            StopCoroutine(_typewriterCoroutine);
            _typewriterCoroutine = null;
        }
    }
    
    /// <summary>
    /// Complete text typing immediately
    /// </summary>
    public void CompleteTyping()
    {
        if (_typewriterCoroutine != null)
        {
            StopCoroutine(_typewriterCoroutine);
            _typewriterCoroutine = null;
            
            if (_dialogueText != null)
            {
                _dialogueText.text = _fullDialogueText;
            }
        }
    }
}
